### ISSUE TITLE
HUGE Inconsistency between logged tokens_per_second_per_GPU and actual wall time and Global Step is Not Monotonically Increasing

### ISSUE THREAD

**Issue by aflah02** [NVIDIA/NeMo#12727]
Hi

I'm training a Llama 3.2 1B style model on 100 B tokens.

This is the wandb log for a run on 8xH100 - [llama32_1b_dclm-SL-2048-PGBS-16-GAS-4-NGPU-8-NNODES-1-TW-PERF | Nemo_Testing – Weights & Biases](https://wandb.ai/aflah/Nemo_Testing/runs/o3pwve8t?nw=nwuseraflah)

The tokens per second is consistently above 70K TPS/GPU.

So in terms of time it should take ~`100*10^9/(70000*8*3600*24)` days which is ~2.1 days

The per gpu batch size is 16 and I have 4 gradient accumulation steps. So each step sees `16*4*8` = 512 samples (since we have 8 GPUs). 512 samples is `512*2048`=1048576 tokens. So each step sees 1048576 tokens . Each train step takes ~1.85 seconds as per the wandb logs. If you use these numbers the token per second estimate in the logs is correct so this is nice.

Now, to consume 100B tokens you will need `100*10^9/1048576` steps i.e. 95367 steps but if we look at the actual time I only did 9.5K steps in 1 day. So I am not sure what's going wrong here but it seems tokens_per_second is reported to be much higher than it actually is? However I might be totally miscalculating something here

Another issue is the logged global step which is non-monotonically increasing while it can never really go down so something is broken there

Code - https://gist.github.com/aflah02/6770d841d121432d7b8d6267d2523f0d

**Comment by aflah02** [comment]
I think the callback is broken - https://github.com/NVIDIA/NeMo/blob/main/nemo/utils/exp_manager.py#L268 as it does not handle tokens seen in the batch correctly in cases of gradient accumulation which gives a false impression of high tokens per second. If we use the global batch size then timer should measure time for all batches before a step instead of on a per batch level.
